bakery-src/scripts/print_customizations.py

```diff
@@ -0,0 +1,82 @@
+import sys
+import shutil
+from pathlib import Path
+
+from lxml import etree
+from lxml.html import XHTMLParser
+
+
+NS_XHTML = "http://www.w3.org/1999/xhtml"
+NS_MAP = {"h": NS_XHTML}
+
+
+def say(msg):
+    print(msg, file=sys.stderr)
+
+
+def clean_toc(tree):
+    toc = tree.xpath('//*[@id="toc"]')[0]
+    print_exclude_class = "no-print"
+    page_query = (
+        '//*[@data-type="page"]['
+        f'   contains(concat(" ", @class, " "), " {print_exclude_class} ")'
+        ']'
+    )
+    chapter_query = (
+        '//*[@data-type="chapter"]['
+        '    not(.//*[@data-type="page" or @data-type="composite-page"])'
+        ']'
+    )
+    unit_query = (
+        '//*[@data-type="unit"]['
+        '    not(.//*[@data-type="chapter" or @data-type="composite-chapter"])'
+        ']'
+    )
+    did_match = False
+    # order is important
+    for query in (page_query, chapter_query, unit_query):
+        elements = tree.xpath(query)
+        for element in elements:
+            el_id = ""
+            data_type = element.attrib.get("data-type")
+            assert data_type, "Element without data type"
+            # units do not have ids
+            if data_type in ("page", "chapter"):
+                # pages have ids, chapters have a title with an id
+                el_id = element.xpath('./@id | .//*[@id]/@id')[0]
+                for tocItem in toc.xpath(
+                    f'.//h:a[@href="#{el_id}"]/parent::h:li', namespaces=NS_MAP
+                ):
+                    container = tocItem.getparent()
+                    container.remove(tocItem)
+                    say(f'Removed link to: "{el_id}"')
+                    if len(list(container.iterchildren())) == 0:
+                        container.getparent().remove(container)
+                        say("Removed container parent")
+            element.getparent().remove(element)
```

Comment:
Won't this trigger for all chapters/units? What am I missing?

Reply:
Oh I get it, the not part of the xpath makes sure it's empty.

Reply:
Wait a sec... units could have just a preface no? So they would match even though they are not empty. Not sure if units with only leaf pages are really relevant though.